### PR TITLE
support custom vjp trait

### PR DESCRIPTION
**Pull request by lxd-cumt**
### PR types

New features
### PR changes

Others
### Description

**add custom_vjp trait to support checking whether an op has custom vjp**

For example, gelu op has custom vjp rules and is added into custom_vjp list, as follows,
```python
PRIM_VJP = ['divide_grad', 'sum_grad']  # vjp list of primitive op
CUSTOM_VJP = ['gelu_grad']  # custom vjp list of composite op
VJP_COMPS = PRIM_VJP + CUSTOM_VJP
```
Therefore,  for gelu op, CustomVjpTrait is automatically added into it, as follows,
```c++
class GeluOp : public ir::Op<GeluOp,paddle::dialect::OpYamlInfoInterface,paddle::dialect::InferMetaInterface,paddle::dialect::VjpInterface,paddle::dialect::CustomVjpTrait>
```
Finally, when calling has_custom_vjp(gelu_op), it will return True, representing that the custom vjp rules for gelu are currently defined.

Pcard-66975

### CODE REVIEW

File: test/prim/new_ir_prim/test_custom_vjp_trait.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    )
+    with paddle.static.program_guard(main_program, start_program):
+        x = paddle.static.data('x', [2, 3, 3], dtype='float32')
+        net = nn.GELU()
+        out = net(x)
+    newir_program = ir.translate_to_new_ir(main_program.desc)
+    return newir_program
+
+
+class TestCustomVjpTrait(unittest.TestCase):
```
Comment (Charles-hit) [inline test/prim/new_ir_prim/test_custom_vjp_trait.py:50]: 单测需要加一个反例

**Comment by Charles-hit** [comment]
PR描述最好可以贴一个生成出来的例子，比如gelu

File: paddle/fluid/ir/dialect/paddle_dialect/trait/custom_vjp.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "paddle/ir/core/op_base.h"
+
+namespace paddle {
+namespace dialect {
+class CustomVjpTrait : public ir::OpTraitBase<CustomVjpTrait> {
```
Comment (cyber-pioneer) [inline paddle/fluid/ir/dialect/paddle_dialect/trait/custom_vjp.h]: explain custom vjp

**Comment by lxd-cumt** [comment]
> PR描述最好可以贴一个生成出来的例子，比如gelu

done，thx

File: paddle/fluid/ir/dialect/paddle_dialect/trait/custom_vjp.h (hunk 2 above)
Comment (lxd-cumt) [reply paddle/fluid/ir/dialect/paddle_dialect/trait/custom_vjp.h]: done, thx

**Review by Charles-hit — DISMISSED** [review]

File: test/prim/new_ir_prim/test_custom_vjp_trait.py (hunk 1 above)
Comment (lxd-cumt) [reply test/prim/new_ir_prim/test_custom_vjp_trait.py:50]: okk，下个PR加入

File: paddle/fluid/pir/dialect/op_generator/op_gen.py [hunk 3]
```diff
@@ -30,6 +32,12 @@
     vjp_interface_implementation_gen_op_list,
 )
 
+# import from paddle/fluid/primitive/code_gen/gen.py
+sys.path.append(
+    str(pathlib.Path(__file__).resolve().parents[3] / 'primitive/codegen')
+)
```
Comment (zhangbo9674) [inline paddle/fluid/pir/dialect/op_generator/op_gen.py:38]: 不可以使用相对路径导入的方式么？

**Review by zhangbo9674 — DISMISSED** [review]

File: paddle/fluid/pir/dialect/op_generator/op_gen.py (hunk 3 above)
Comment (lxd-cumt) [reply paddle/fluid/pir/dialect/op_generator/op_gen.py:38]: 当前缺少__init__.py，使用相对路径的话需要新增才可以；
python代码规范里面也提到：
```python
Absolute imports are recommended, as they are usually more readable and tend to be better behaved 
(or at least give better error messages) if the import system is incorrectly configured .
```
所以使用了绝对路径。

**Review by cyber-pioneer — APPROVED** [review]

**Pull request merged**